sui_core/src/transaction_input_checker.rs

```diff
@@ -3,22 +3,107 @@
 
 use std::collections::HashSet;
 
+use serde::{Deserialize, Serialize};
 use sui_types::{
-    base_types::{ObjectRef, SequenceNumber, SuiAddress},
+    base_types::{ObjectID, ObjectRef, SequenceNumber, SuiAddress},
     error::{SuiError, SuiResult},
     fp_ensure,
-    messages::{InputObjectKind, SingleTransactionKind, TransactionData},
+    gas::{self, SuiGasStatus},
+    messages::{InputObjectKind, SingleTransactionKind, TransactionData, TransactionEnvelope},
     object::{Object, Owner},
 };
-use tracing::debug;
+use tracing::{debug, instrument};
+
+use crate::authority::SuiDataStore;
+
+#[instrument(level = "trace", skip_all)]
+pub async fn check_transaction_input<const A: bool, S, T>(
+    store: &SuiDataStore<A, S>,
+    transaction: &TransactionEnvelope<T>,
+) -> Result<(SuiGasStatus<'static>, Vec<(InputObjectKind, Object)>), SuiError>
+where
+    S: Eq + Serialize + for<'de> Deserialize<'de>,
+{
+    let (gas_object, mut gas_status) = check_gas(
+        store,
+        transaction.gas_payment_object_ref().0,
+        transaction.data.gas_budget,
+    )
+    .await?;
+
+    let objects_by_kind =
+        check_locks(store, &transaction.data, gas_object, &mut gas_status).await?;
+
+    if transaction.contains_shared_object() {
+        // It's important that we do this here to make sure there is enough
+        // gas to cover shared objects, before we lock all objects.
+        gas_status.charge_consensus()?;
```

Comment:
Should this down the line take more params, like the number of shared objects etc?

Reply:
Does the number of shared objects used affect the cost of consensus?